graph_worker.cpp

```diff
@@ -19,37 +29,74 @@ void GraphWorker::startWorkers(Graph *_graph, BufferTree *_bf) {
 	}
 }
 
-void GraphWorker::stopWorkers() {
+void GraphWorker::stop_workers() {
 	shutdown = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+	pause_condition.notify_all();      // tell any paused threads to continue and exit
 	for (int i = 0; i < num_groups; i++) {
 		delete workers[i];
 	}
 	delete workers;
 }
 
+void GraphWorker::pause_workers() {
+	paused = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+
+	// wait until all GraphWorkers are paused
+	std::unique_lock<std::mutex> lk(pause_lock);
+	pause_condition.wait(lk, []{
+		for (int i = 0; i < num_groups; i++)
+			if (!workers[i]->get_thr_paused()) return false;
+		return true;
+	});
+	lk.unlock();
+}
+
+void GraphWorker::unpause_workers() {
+	workers[0]->bf->bypass_wait(false); // buffer-tree operations should block when necessary
+	paused = false;
+	pause_condition.notify_all();       // tell all paused workers to get back to work
+}
+
+/***********************************************
+ ************** GraphWorker class **************
+ ***********************************************/
 GraphWorker::GraphWorker(int _id, Graph *_graph, BufferTree *_bf) :
-  id(_id), graph(_graph), bf(_bf), thr(startWorker, this) {
+  id(_id), graph(_graph), bf(_bf), thr(start_worker, this) {
 }
 
 GraphWorker::~GraphWorker() {
-	if(id == 0)
-		bf->bypass_wait(); // to avoid race condition on shutdown. Tell threads to recheck wait
-	
-	// when a graph worker is deleted we must wait for the associated
-	// thread to finish its work. See stopWorkers and doWork
+	// join the GraphWorker thread to reclaim resources
 	thr.join();
 }
 
-void GraphWorker::doWork() {
+void GraphWorker::do_work() {
 	data_ret_t data;
 	while(true) {
-		// call get_data which will handle waiting on the queue
-		// and will enforce locking.
-		bool valid = bf->get_data(data);
-
-		if (valid)
-			graph->batch_update(data.first, data.second);
-		else if(shutdown)
+		if(shutdown)
 			return;
+		pause_lock.lock();
```

Comment:
an atomic variable is best suited for this semantic

Reply:
Yes good point. I'll make the change